src/libraries/System.IO.FileSystem/tests/Directory/Copy.cs

```diff
@@ -0,0 +1,216 @@
+            Directory.Copy(sourceDirName, destDirName, true, false, cancellationToken);
+
+            // Validate cancellation
+            Assert.False(cancellationToken.IsCancellationRequested);
+
+            // Validate copied files
+            string destTopLevelFilePath = Path.Combine(destDirName, "top-level.txt");
+            string destSecondLevelDirectoryPath = Path.Combine(destDirName, "subdir");
+            string destThirdLevelFilePath = Path.Combine(destSecondLevelDirectoryPath, "test.txt");
+
+            Assert.True(File.Exists(destTopLevelFilePath));
+            Assert.True(File.Exists(destThirdLevelFilePath));
+        }
+
+        [Fact]
+        public void Copy_NonRecursive()
+        {
+            const string sourceDirName = $"Test_source_{nameof(Copy_NonRecursive)}";
+            const string destDirName = $"Test_target_{nameof(Copy_NonRecursive)}";
+            CancellationToken cancellationToken = new();
+
+            // Prerequisites
+            // Create source folder
+            if (Directory.Exists(sourceDirName))
+            {
+                Directory.Delete(sourceDirName, true);
+            }
+            Directory.CreateDirectory(sourceDirName);
+
+            // Create destination folder
+            if (Directory.Exists(destDirName))
+            {
+                Directory.Delete(destDirName, true);
+            }
+            Directory.CreateDirectory(destDirName);
+
+            string sourceTopLevelFilePath = Path.Combine(sourceDirName, "top-level.txt");
+            using (StreamWriter sourceTopLevelFileStream = File.CreateText(sourceTopLevelFilePath))
+            {
+                sourceTopLevelFileStream.WriteLine("Test");
+            }
+
+            string sourceSecondLevelDirectoryPath = Path.Combine(sourceDirName, "subdir");
+            Directory.CreateDirectory(sourceSecondLevelDirectoryPath);
+
+            string sourceThirdLevelFilePath = Path.Combine(sourceSecondLevelDirectoryPath, "test.txt");
+            using (StreamWriter sourceThirdLevelFileStream = File.CreateText(sourceThirdLevelFilePath))
+            {
+                sourceThirdLevelFileStream.WriteLine("Test");
+            }
+
+            // Do file system opreation
+            Directory.Copy(sourceDirName, destDirName, false, false, cancellationToken);
+
+            // Validate cancellation
+            Assert.False(cancellationToken.IsCancellationRequested);
+
+            // Validate copied files
+            string destTopLevelFilePath = Path.Combine(destDirName, "top-level.txt");
+            string destSecondLevelDirectoryPath = Path.Combine(destDirName, "subdir");
+            string destThirdLevelFilePath = Path.Combine(destSecondLevelDirectoryPath, "test.txt");
+
+            Assert.True(File.Exists(destTopLevelFilePath));
+
+            // False because we only want to copy top level files (= non-recursive)
+            Assert.False(File.Exists(destThirdLevelFilePath));
+        }
```

Comment:
You need to add tests for read and write errors, and for the simpler `Directory.Copy` overload.

Reply:
In my last commit, I added these. However, the two tests that check for it will fail. 

I don't know exactly why but I guess it's still due to the problem that I don't know exactly how to clearly distinguish the read and write failures....

Maybe you can find another approach.

Feel free to cancel the CI, I'm sure it won't pass.